### PR TITLE
Modularity

### LINKED ISSUE #248: Breaking things up
As fixes and features are added to this library things can quickly become bloated.
This library is still small but to aid future contributors and preserve code manageability, I propose restructuring the library now rather than later. I recommend modeling after `requests`.

For a quick comparison of what I aim to avoid, consider another extremely popular request library.

Bloated: https://github.com/request/request/blob/master/request.js
It took the poor guy quite a bit of time to refactor the library: https://gist.github.com/simov/552bee6ce6001205393e
It hasn't happened yet: https://github.com/request/request/issues/1982 :cry: 

This is what I imagine the `__init__.py` would initially look like.
```py
from .adapters import CloudflareAdapter, DEFAULT_CIPHERS
from .exceptions import CloudflareError, CloudflareCaptchaError
from .api import request, get, head, post, patch, put, delete, options

from .scraper import CloudflareScraper, user_agents, \
    DEFAULT_USER_AGENT, DEFAULT_HEADERS  

__version__ = "2.0.4"

create_scraper = CloudflareScraper.create_scraper
get_tokens = CloudflareScraper.get_tokens
get_cookie_string = CloudflareScraper.get_cookie_string
```

Modeled after `requests`, you could perform the following:
```py
import cfscrape
cfscrape.get(url)
```

This idea also makes for better tests.

### PR DESCRIPTION
**Pull request by pro-src**
I used two commits in an attempt to make this less painful when viewing the diff in some cases. I highly recommend viewing the diff after the rename: https://github.com/Anorov/cloudflare-scrape/compare/f887510..f67bd4f

This adds `cfscrape/api.py` as discussed, more tests, and increases code coverage: https://coveralls.io/jobs/48653621

Closes #248 <strike>although I would like to see a `cfscrape/sandbox.py` eventually.</strike>
I created `cfscrape/challenge.py` for that purpose. I'm not sure a sandbox class would be that useful now.

### CODE REVIEW

File: cfscrape/api.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+
+from .scraper import CloudflareScraper
+
+
+scraper = None
```
Comment (pro-src) [inline cfscrape/api.py]: If preferable, we could create the scraper when imported instead of deferring until the API is actually used.

Comment (pro-src) [reply cfscrape/api.py]: i.e. The entire contents of this file could be replaced with:
```py
# -*- coding: utf-8 -*-

import sys
from .scraper import CloudflareScraper


sys.modules[__name__] = CloudflareScraper()
```

File: cfscrape/scraper.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        if self.is_cloudflare_captcha_challenge(resp):
+            self.handle_captcha_challenge(resp, url)
+
+        # Check if Cloudflare anti-bot "I'm Under Attack Mode" is enabled
+        if self.is_cloudflare_iuam_challenge(resp):
+            resp = self.solve_cf_challenge(resp, **kwargs)
+
+        return resp
+
+    def cloudflare_is_bypassed(self, url, resp=None):
```
Comment (pro-src) [inline cfscrape/scraper.py]: I think this got added by @lukele PR, probably can remove it now.

File: cfscrape/scraper.py [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
 
-from base64 import b64encode
 from collections import OrderedDict
 
 from requests.sessions import Session
 from requests.compat import urlparse, urlunparse
 
 from .adapters import CloudflareAdapter
 from .exceptions import CloudflareCaptchaError
+from .challenges import CloudflareChallenge
```
Comment (pro-src) [inline cfscrape/scraper.py]: @Anorov 

This should make it easy to add another challenge type i.e. `CaptchaChallenge` which could be the base of `CloudflareCaptchaChallenge` and possibly a [hCaptcha](https://hcaptcha.com) class. Or perhaps just a `BaseChallenge`...

Parsing the siteKey and form values from the captcha page so a user can solve and submit would be an awesome feature.

**Comment by Anorov** [comment]
Will review this later tonight, or tomorrow night.

File: cfscrape/scraper.py (hunk 3 above)
Comment (Anorov) [reply cfscrape/scraper.py]: What scenarios did you have in mind for supporting alternative captchas like hCaptcha? As far as I know, Cloudflare only uses reCAPTCHA, and I don't plan on adding support for services other than Cloudflare. Maybe I'm misunderstanding something.

Comment (pro-src) [reply cfscrape/scraper.py]: No, you're right. I really meant to just note that adding other challenge types will be easier.

Aside from that, I was thinking that there wouldn't be much difference between the two but I really don't know. I know that https://github.com/privacypass/challenge-bypass-extension supports both and the extension itself seems to be supported/encouraged by Cloudflare. I don't know that I've ever encountered a hCaptcha and didn't have it in mind to personally add support for it.

Comment (pro-src) [reply cfscrape/scraper.py]: Example class: https://gist.github.com/pro-src/de7d48eb20839cb35febf55a6758e965

**Comment by Anorov** [comment]
Haven't had much time lately, unfortunately. Will get to it within a few days.

**Comment by pro-src** [comment]
Thx for the update! No worries, no rush. In the meantime, does anybody else want to leave a review?

Past contributors. @lukele, @lukastribus maybe?

Any comments from downstream projects would be especially welcomed.

**Comment by pro-src** [comment]
Bump. :)

**Comment by pro-src** [comment]
I'm closing this for now. Feel free to reopen at a later time, branch will still be there.

**Pull request closed without merging**